### DISCUSSION TITLE
How to add custom checkbox and radiobutton renderer?

### DISCUSSION THREAD

**Discussion by ApayRus** [remarkjs/remark#619] [Q&A] [answered]
Sorry if my question is stupid. I was using marked.js and now it's hard to switch mindset into remark :) 
Please take a look on the question and provide me with a quick guide, like a couple of links to docs that I needed etc... 

I need to render `task list`: `- [x] list item 1` in a tricky way. Not like in GFM spec. 

1) `checked` should be set with any non-space symbol, because [xX] is hard to type in any non-Latin language (everytime you should switch keyboards witch is pain especially when you write exams sheets with thousands of lines). 

2) completely rewrite <input type="checkbox"... /> into custom tag: `<checkbox checked />` . 
```
- [v] variant 1
- [ ] variant 2
- [v] variant 3
- [ ] variant 4
```
should be rendered to 
```jsx
<ul class="quiz">
  <li> <checkbox checked /> variant 1 </li>
  <li> <checkbox /> variant 2 </li>
  <li> <checkbox checked /> variant 3 </li>
  <li> <checkbox /> variant 4 </li>
</ul> 
```
3) add radio-button in place of round brackets:

```
- ( ) variant 1
- (v) variant 2
- ( ) variant 3
```
should be rendered to 
```jsx
  <ul class="quiz">
  <li> <roundbutton /> variant 1 </li>
  <li><roundbutton checked />  variant 2 </li>
  <li> <roundbutton />  variant 3 </li>
</ul> 
```
4) Can I extract an index of `<ul>` , its order in markdown text? For render it: `<ul class="quiz" id="1" >`, 2, 3,... ?

**Comment by wooorm** [comment, marked as answer]
> Not like in GFM spec.

remark is standards based. You might want to create an extension, but I discourage it. It makes your markdown brittle: the next time you change markdown parsers, the same problem will come up

1. Can you copy-paste them?
2. remark is AST based: objects (called nodes) represent constructs (so, tasklists). First there’s a markdown tree (many nodes), where `ListItem` nodes have a `checked` field, set to a boolean. That gets transformed to an HTML tree (hast / rehype), where there are nodes representing HTML elements (so there’s an `Element` node with a `tagName` set to `input`). [You can create a plugin, traverse the tree, and change HTML elements](https://unifiedjs.com/learn/). I strongly suggest against `checkbox` and `roundbutton` elements: those are incorrect HTML.
3. See the earlier note on extensions
4. While traversing the tree, you have access to the `index` a `listItem` (in markdown/mdast/remark) or `li` element (in HTML/hast/rehype) is at

---

Why do you need `roundbutton` and `checkbox` elements? Maybe this is an [XY problem](https://xyproblem.info)

**Reply by ApayRus** [reply]
Thank you for quick answer! It is hard to absorb it... but I try...

> Why do you need `roundbutton` and `checkbox` elements? Maybe this is an [XY problem](https://xyproblem.info)
I'm working on mobile app, and all my custom `tags` converts to react-native elements. 

I do app for quiz/poll/test sheets for students. Teacher writes in markdown questions, variants and checks correct answers. Checkbox means many possible correct answers, roundbutton -- only one correct answer. Then I compile markdown into html with custom tags -- as I've written above. my app renders this html into react-native interface. Students don's see correct answers until they press button `check answer`. Then they get rates.

**Reply by wooorm** [reply]
`Checkbox means many possible correct answers, roundbutton -- only one correct answer`

Can you detect this? As in, if there is a single checked list item, it’s one correct answer, if there are multiple, it’s multiple choice?

**Reply by ApayRus** [reply]
> Can you detect this? As in, if there is a single checked list item, it’s one correct answer, if there are multiple, it’s multiple choice?

not exactly... ckeckbox/radiobutton -- its not only about count of correct answers, but also ability to interact with group of elements. Group of radiobuttons don't get you check less or more than 1 answer. Group of ckeckboxes get you ckeck 0-countOfVariants, -- 0 and 1 also possible. It's completly different logic, and than I thought to separate them in markdown `[ ] / ( ) ` otherwise i should write somewhere : "checkboxes group" or "radiobutton group" and it will be a noise in text... 

> 1. Can you copy-paste them?

sorry i didn't understand what you meen.

**Reply by wooorm** [reply]
> sorry i didn't understand what you meen.

Write `[x]` once by switching keyboards, then copy and paste them if you need more! It’s not ideal of course, but it would prevent breaking from GFM (which is also not ideal)

**Reply by ChristianMurphy** [reply]
Another option would be to change the syntax a bit and leverage [directives](https://github.com/remarkjs/remark-directive) which are [semi-standard](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444) and more expressive than checkbox lists.
Directives to multiple choice could look something like:

```markdown
:::multiple-choice[select-one]

::option[some text]
::option[some different text]{correct=true}

:::

---

:::multiple-choice[select-multiple]

::option[some text]
::option[some different text]{correct=true}

:::
```
https://astexplorer.net/#/gist/b9a1eb6983f0d8151f7af796aa6c9afd/11c93497a79fc349f9dc0a86bfe9be9a7fa6f559

**Reply by ApayRus** [reply]
> Another option would be to change the syntax a bit and leverage [directives](https://github.com/remarkjs/remark-directive) which are [semi-standard](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444) and more expressive than checkbox lists.

@ChristianMurphy , thank you , I didn't know about it and it's good to know. But I don't like it... 
Is my idea crime? about adding `- ( ) ` for radiobutton ?

**Reply by ChristianMurphy** [reply]
The choice of syntax is up to you.
I'd echo the consideration @wooorm raised, that creating and maintaining a parser can be a lot of time and effort.
If it's the syntax you really want, that effort of creating the parser may be worth while.
If you're looking for something quick to setup and low maintenance, any of the alternative syntax possibilities noted above could be a good option.

**Reply by ApayRus** [reply]
@wooorm , when you told "extention" , did you mean "plugin" ? `docs/plugin` section? 
Also i didn't understand connection between `unified` and `remark`, i tried to get it couple of times but it is too hard.  is unified == remark? Should I learn unified before use remark? 
 when i wrote a question, i expected that someone share with me some code like:
```js
marked().use({plugin: (text) => text.replace (/- ( ) /, `<radiobutton />` } ) 
```
But now I should read a lot of texts about philosophy of `unified-community` on other site, where even no word 'remark' (but I've found it with ctrl+f)... it looks like 'remarl' ))
![Screenshot from 2021-01-27 21-30-13](https://user-images.githubusercontent.com/1222611/106036873-eecba480-60e6-11eb-9943-d7c772d034f5.png)

About fear of using custom tags, I am using [react-native-render-html ](https://github.com/meliorence/react-native-render-html). And there are half of documentation about custom tags. I didn't think it is bad using them, until @wooorm scared me. 

Pardon my nagging. Can you provide me with peace of code examples? I'm going through the links but don't see examples, which I expected on Readme page. 

To avoid XY problem...  (I didn't know about it, thank you again!) and wide perspective... 
I made a mobile app template for language learning apps like Duolingo etc, that completely is generated from markdown text and some assets (images, audios, fonts).  Drawer menu, Cower, languages switcher, exercises, multilanguage support -- everything generated for simplest markdown as possible -- so clear as a note in plain text... I'm not that guy who reinvent the wheel. But I stuck with that idea of quiz :(  But quiz is important part of any edu app. 

![universal language learning app](https://i.imgur.com/Fbo8YgL.gif)
![exercises](https://i.imgur.com/42owMhq.gif)

**Reply by ChristianMurphy** [reply]
There are guides for Unified as a whole, using plugins, and creating plugins in https://unifiedjs.com/learn

> Also i didn't understand connection between unified and remark

Again https://unifiedjs.com/learn is a good resource.
At a high level, `unified` is a utility for working with Abstract Syntax trees.
`remark` (markdown), `rehype` (html), `retext` (natural language), `mdx` (markdown + jsx), and `redot` (graphviz) are parsers for specific languages that can be used with `unified`.

> Can you provide me with [...] code examples? I'm going through the links but don't see examples, which I expected on Readme page.

Where are you looking?
Unified and the surrounding ecosystem has 300+ packages/utilities, each readme usually includes usage example.

> But I stuck with that idea of quiz

There are a lot of options for quiz.
If you want something which is more standardized, https://en.wikipedia.org/wiki/QTI is a solid and widely supported standard for quizzes and tests.
It's in XML and https://github.com/syntax-tree/xast (another unified parser utility) can be used to interact with QTI files.

If you don't like XML, then options like [directives for custom elements in markdown](https://github.com/remarkjs/remark-directive), [mdx for custom elements in JSX](https://mdxjs.com), or using code blocks to embed another language inside markdown (here's an example with [mermaid embedded in markdown](https://github.com/temando/remark-mermaid)).
Though choosing a custom syntax for quizzes would be a bit of "reinventing the wheel". :slightly_smiling_face: :man_shrugging:

**Reply by ApayRus** [reply]
@ChristianMurphy . Thank you. 

 > reinventing the wheel

I mean, I made more things, the whole app, without adding anything to Markdown. And I believe that my format for quiz is good and justified. 

> Where are you looking?

On remark Readme.md, then in docs/plugins, everywhere links to other pages. That demotivated me several times.

Now I've read carefully  https://unifiedjs.com/learn/guide/introduction-to-unified/ and understand better and even made mindmap: 
![Unified js xmind - Map 1](https://user-images.githubusercontent.com/1222611/106069945-94e1d380-6114-11eb-8da0-aa15d1a9999b.png). 

Thank you all again. Hope we will collaborate in the future. I like your ideas so much. It is what I was thinking about, but more deeply and I want to follow you. It is exactly what I was looking for: 

> Where unified really shines is when you want to go further than one single task. For example, when you want to enforce format rules, check spelling, generate a table of contents, and (potentially) much more: that’s when to opt for unified.

Ok, I'll study all theory on  main page. Sorry for my hurry. Bye.
